File: ox-alpha_dungeon_crawler/src/entities/Hunter.js
// Hunter.js — HUNTER buff companion: follow, LOS-targeted beam (§11)
import * as THREE from 'three';
import { HUNTER } from '../core/Constants.js';

export default class Hunter {
  constructor(scene) {
    this.pos = new THREE.Vector3(0, 0, 0);
    this.attackTimer = 0;
    this.beamFlash = 0;
    const mat = new THREE.MeshStandardMaterial({
      color: 0x88aaff, emissive: 0x4466cc, emissiveIntensity: 1.2,
      transparent: true, opacity: 0.85
    });
    this.group = new THREE.Group();
    const body = new THREE.Mesh(new THREE.ConeGeometry(0.3, 1.0, 8), mat);
    body.position.y = 1.2;
    const head = new THREE.Mesh(new THREE.SphereGeometry(0.16, 8, 6), mat);
    head.position.y = 1.9;
    this.group.add(body, head);
    scene.add(this.group);
    this._mat = mat;
  }

  update(dt, playerPos, enemies, losFn, onHitEnemy, souls) {
    // follow at 6.5 u/s keeping 2.5 u
    const dx = playerPos.x - this.pos.x, dz = playerPos.z - this.pos.z;
    const d = Math.hypot(dx, dz);
    if (d > HUNTER.KEEP_DIST) {
      const step = Math.min(HUNTER.FOLLOW_SPEED * dt, d - HUNTER.KEEP_DIST);
      this.pos.x += (dx / d) * step;
      this.pos.z += (dz / d) * step;
    }
    this.group.position.set(this.pos.x, 0, this.pos.z);

    // attack nearest VISIBLE enemy within 7 u; interval 1.0/clamp(souls/100, .25, 5)
    this.attackTimer -= dt;
    if (this.beamFlash > 0) this.beamFlash -= dt;
    if (this.attackTimer <= 0) {
      let best = null, bestD = HUNTER.ATTACK_RANGE;
      for (const e of enemies) {
        if (!e.alive || e.state === 'DEAD' || e.frozen) continue;
        const dd = Math.hypot(e.pos.x - this.pos.x, e.pos.z - this.pos.z);
        if (dd < bestD && losFn(this.pos.x, this.pos.z, e.pos.x, e.pos.z)) { bestD = dd; best = e; }
      }
      if (best) {
        onHitEnemy(best, HUNTER.BEAM_DMG);
        this.beamFlash = HUNTER.BEAM_FLASH;
        this.attackTimer = 1.0 / Math.min(5, Math.max(0.25, souls / 100));
      }
    }
  }

  dispose(scene) {
    scene.remove(this.group);
    this.group.traverse(o => {
      if (o.geometry) o.geometry.dispose();
      if (o.material) o.material.dispose();
    });
  }
}

File: c++_dungeon_crawler/src/core/hunter.cpp
// dc/hunter.cpp — Friendly companion: follow + beam at mobs and breakables.
#include "dc/hunter.hpp"

#include <algorithm>
#include <cmath>

namespace dc {

void Hunter::update(double dt, const Vec2& playerPos,
                    const std::vector<Enemy*>& enemies,
                    const std::vector<Breakable*>& breakables,
                    const std::function<bool(const Vec2&, const Vec2&)>& los,
                    const std::function<void(Enemy*)>& onHitEnemy,
                    const std::function<void(Breakable*)>& onHitBreakable,
                    double souls) {
  // follow at FOLLOW_SPEED keeping KEEP_DIST (behind the player)
  const double dx = playerPos.x - pos.x, dz = playerPos.z - pos.z;
  const double d = std::hypot(dx, dz);
  if (d > hunter::kKeepDist) {
    const double step = std::min(hunter::kFollowSpeed * dt, d - hunter::kKeepDist);
    pos.x += (dx / d) * step;
    pos.z += (dz / d) * step;
  }

  attackTimer -= dt;
  if (beamFlash > 0) beamFlash -= dt;
  if (beamFlash <= 0) { hasBeam = false; beamTargets.clear(); }

  if (attackTimer <= 0) {
    // Collect eligible targets: enemies + breakables, nearest first.
    struct Cand { double d; bool isEnemy; Enemy* enemy = nullptr; Breakable* br = nullptr; };
    std::vector<Cand> cands;

    // Enemies (alive, not frozen, LOS check)
    for (Enemy* e : enemies) {
      if (!e->alive() || e->frozen) continue;
      const double dd = std::hypot(e->pos.x - pos.x, e->pos.z - pos.z);
      if (dd < hunter::kAttackRange && los(pos, e->pos))
        cands.push_back({dd, true, e, nullptr});
    }

    // Breakables (alive, LOS check)
    for (Breakable* br : breakables) {
      if (!br->alive) continue;
      const double dd = std::hypot(br->pos.x - pos.x, br->pos.z - pos.z);
      if (dd < hunter::kAttackRange && los(pos, br->pos))
        cands.push_back({dd, false, nullptr, br});
    }

    if (cands.empty()) return;
    std::sort(cands.begin(), cands.end(),
              [](const Cand& a, const Cand& b) { return a.d < b.d; });
    beamTargets.clear();
    hasBeam = true;
    beamFlash = hunter::kBeamFlash;
    for (size_t i = 0; i < cands.size() && i < (size_t)hunter::kMaxBeamTargets; i++) {
      beamTargets.push_back(cands[i].isEnemy ? cands[i].enemy->pos : cands[i].br->pos);
      if (cands[i].isEnemy && cands[i].enemy)
        onHitEnemy(cands[i].enemy);
      else if (cands[i].br)
        onHitBreakable(cands[i].br);
    }
    attackTimer = 1.0 / std::min(5.0, std::max(0.25, souls / 100.0));
  }
}

} // namespace dc
